src/disc_bsd.c

```diff
@@ -23,29 +23,37 @@
 
 --------------------------------------------------------------------------- */
 
+#include <assert.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
-#include <sys/types.h>
 #include <sys/cdio.h>
 #include <sys/ioctl.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#if !defined(__FreeBSD__)
+#include <util.h>
+#endif
 
 #include "discid/discid_private.h"
 #include "unix.h"
 
-int mb_disc_unix_read_toc_header(int fd, mb_disc_toc *toc) {
-	struct ioc_toc_header th;
-
-	int ret = ioctl(fd, CDIOREADTOCHEADER, &th);
+#define MAX_DEV_LEN 15
 
-	if ( ret < 0 )
-		return 0; /* error */
-
-	toc->first_track_num = th.starting_track;
-	toc->last_track_num = th.ending_track;
-
-	return 1;
+static int get_device(int n, char* device_name, size_t device_name_length) {
+#if !defined(__FreeBSD__) /* /dev/rcdNX, where X is the letter for the raw partition */
+	snprintf(device_name, device_name_length, "/dev/rcd%d%c", n - 1, 'a' + getrawpartition());
+#else /* on FreeBSD it's just /dev/cdN */
+	snprintf(device_name, device_name_length, "/dev/cd%d", n - 1);
+#endif
+	return mb_disc_unix_exists(device_name);
 }
 
+/* not actually used for *BSD, but needed by unix.c */
```

Comment:
Because the unix implementation insists on reading each toc entry separately. All BSDs have a read toc entries ioctl allowing them all to be read in one go; it seemed silly to issue N ioctls in that case.
I suppose I could do that in the read_toc_header function, and leave the entry one as a no-op. That's still several unecessary function calls, but a minimum of ioctls.